lapack/testlapack/dlagv2.go

```diff
@@ -0,0 +1,182 @@
+type Dlagv2er interface {
+	Dlagv2(a []float64, lda int, b []float64, ldb int, alphar, alphai, beta []float64) (csl, snl, csr, snr float64)
+}
+
+func Dlagv2Test(t *testing.T, impl Dlagv2er) {
+	rnd := rand.New(rand.NewSource(1))
+	const ldExtra = 0
+	for _, lda := range []int{2, 2 + ldExtra} {
+		for _, ldb := range []int{2, 2 + ldExtra} {
+			for aKind := 9; aKind <= 20; aKind++ {
+				for bKind := 9; bKind <= 20; bKind++ {
+					dlagv2Test(t, impl, rnd, lda, ldb, aKind, bKind)
+				}
+			}
+		}
+	}
+}
+
+func dlagv2Test(t *testing.T, impl Dlagv2er, rnd *rand.Rand, lda, ldb int, aKind, bKind int) {
+	const tol = 1e-14
+
+	a := makeDlag2TestMatrix(rnd, lda, aKind)
+	b := makeDlag2TestMatrix(rnd, ldb, bKind)
+	b.Data[b.Stride] = math.NaN() // b is lower triangular.
+	// a.Data[a.Stride] = 0
+	// Enforce positive, non-zero diagonal.
+	b.Data[0] = math.Max(safmin, math.Abs(b.Data[0]))
+	b.Data[b.Stride+1] = math.Max(safmin, math.Abs(b.Data[b.Stride+1]))
+
+	if b.Data[0] < b.Data[b.Stride+1] {
+		// Enforce descending order.
+		b.Data[0], b.Data[b.Stride+1] = b.Data[b.Stride+1], b.Data[0]
+	}
+
+	aCopy := cloneGeneral(a)
+	bCopy := cloneGeneral(b)
+	bCopy.Data[bCopy.Stride] = 0 // zero out NaN for later comparisons.
+	var alphar, alphai, beta [2]float64
+	csl, snl, csr, snr := impl.Dlagv2(a.Data, a.Stride, b.Data, b.Stride, alphar[:], alphai[:], beta[:])
+	beta1, beta2 := beta[0], beta[1]
+	wi1, wi2 := alphai[0], alphai[1]
+	wr1, wr2 := alphar[0], alphar[1]
+	if beta1 == 0 {
+		beta1 = 1
+		// wi1, wr1 = wi1/beta1, wr1/beta1
+	}
+	if beta2 == 0 {
+		beta2 = 1
+		// wi2, wr2 = wi2/beta2, wr2/beta2
+	}
+	// Generate r1, r2 rotation matrices.
+	// r1 = [ CSL  SNL; -SNL CSL]
+	r1 := blas64.General{
+		Data: []float64{csl, snl, -snl, csl},
+		Rows: 2, Cols: 2, Stride: 2,
+	}
+	// r2 = [ CSR -SNR; SNR CSR]
+	r2 := blas64.General{
+		Data: []float64{csr, -snr, snr, csr},
+		Rows: 2, Cols: 2, Stride: 2,
+	}
+	isComplexEig := wi1 != 0
+	name := fmt.Sprintf("cmplx=%v,lda=%d,ldb=%d,aKind=%d,bKind=%d", isComplexEig, lda, ldb, aKind, bKind)
+	// aOrigStr := sprint2x2Block("A", aCopy.Data, aCopy.Stride, 4)
+	// bOrigStr := sprint2x2Block("B", bCopy.Data, bCopy.Stride, 4)
+
+	aStr := sprint2x2Block("As", a.Data, a.Stride, 4)
+	bStr := sprint2x2Block("Bs", b.Data, b.Stride, 4)
+
+	if wi1 < 0 {
+		t.Fatalf("%s: negative wi1; wi1=%g,wi2=%g,\n%s\n%s", name, wi1, wi2, aStr, bStr)
+		return
+	}
+	if isComplexEig && wi2 != -wi1 {
+		t.Fatalf("%s: wi1 != -wi2; wi1=%g,wi2=%g,\n%s\n%s", name, wi1, wi2, aStr, bStr)
+		return
+	}
+	if math.Abs(b.Data[b.Stride]) > tol {
+		t.Fatalf("%s: expected b to remain upper triangular:\n%s", name, bStr)
+		return
+	}
+	if isComplexEig && b.Data[0] < b.Data[b.Stride+1] {
+		t.Errorf("%s: expected b diagonal elements to be in descending order:\n%s", name, bStr)
+	}
+
+	if !isComplexEig && !isSchurCanonicalGeneral(a, tol) {
+		t.Fatalf("%s: a is not Schur canonical:\n%s", name, aStr)
+		return
+	}
+
+	if !isSchurCanonicalGeneral(b, tol) {
```

Comment:
B is expected to be upper triangular or diagonal, which is already tested for, so this test seems superfluous.